url1cbaffb197_foxking0416_AlgorithmTraining_tgz-pJ8-algorithm_leetcode_MainLeetJ8/src/datastructure/binarytree/BinaryTree.java

```diff
@@ -0,0 +1,1106 @@
+			}
+			if(rNode.getRightLeafNode() != null){
+				nextLevel.add(rNode.getRightLeafNode());
+			}
+
+			
+			if(currentLevel.isEmpty() == false){
+				rNode = currentLevel.poll();
+			}
+			else{
+				
+				ArrayList<Integer> storeArray = new ArrayList<>(tempArray);
+				/*System.out.print("[");
+				for(Integer i : tempArray){
+					storeArray.add(i);
+					System.out.print(i + ",");
+				}
+				System.out.print("]");*/				
+				array.add(storeArray);
+				tempArray.clear();
+				
+				if(nextLevel.isEmpty() == false){
+
+					for(TreeNode n : nextLevel) {
+						currentLevel.add(n);
+				    }
+					
+					nextLevel.clear();
+					rNode = currentLevel.poll();
+				}
+				else{
+					break;
+				}
+			}
+
+		}
+	}
+	
+	//Zigzag-Order Traverse method
+	public void zigzagLevelOrderTraverse(TreeNode rNode){
+
+		Stack<TreeNode> currentLevel = new Stack<TreeNode>();
+		Stack<TreeNode> nextLevel = new Stack<TreeNode>();
+		boolean leftToRight = true;
+		
+		while(rNode != null){
+			System.out.print(rNode.val + ",");
+			
+			if(leftToRight){
+				if(rNode.getLeftLeafNode() != null){
+					nextLevel.push(rNode.getLeftLeafNode());
+				}
+				if(rNode.getRightLeafNode() != null){
+					nextLevel.push(rNode.getRightLeafNode());
+				}
+			}else {
+				if(rNode.getRightLeafNode() != null){
+					nextLevel.push(rNode.getRightLeafNode());
+				}
+				if(rNode.getLeftLeafNode() != null){
+					nextLevel.push(rNode.getLeftLeafNode());
+				}
+			}
+
+
+			
+			if(currentLevel.isEmpty() == false){
+				rNode = currentLevel.pop();
+			}
+			else{
+				if(nextLevel.isEmpty() == false){
+					currentLevel = (Stack<TreeNode>)nextLevel.clone();// = nextLevel;
+					
+					nextLevel.clear();
+					leftToRight = !leftToRight;
+					rNode = currentLevel.pop();
+				}
+				else{
+					break;
+				}
+			}
+
+		}
+	}
+
+	
+	public TreeNode inorderSucc(TreeNode n){
+		if(n == null)
+			return null;
+		
+		if(n.getRightLeafNode() != null){
+			return findTheMostLeftNode(n.getRightLeafNode());
+		}
+		else{
+			TreeNode parentNode = findParentNode(n, this.rootNode);
+			
+			while (parentNode != null) {
+				if( n.val < parentNode.val){
+					break;
+				}
+				n = parentNode;
+				parentNode = findParentNode(n, this.rootNode);
+			}
+			
+			
+			return parentNode;
+		}
+		
+		
+		
+	}
+	
+	public TreeNode findTheMostLeftNode(TreeNode n){
+		
+		if(n.getLeftLeafNode() == null)
+			return n;
+		else 
+			return findTheMostLeftNode(n.getLeftLeafNode());
+	}
+	
+	public boolean cover(TreeNode root, TreeNode n){
+		if(root.equals(n))
+			return true;	
+		else{
+			boolean leftCover = false;
+			boolean rightCover = false;
+			if(root.getLeftLeafNode() != null)
+				leftCover = cover(root.getLeftLeafNode(), n);
+			if(root.getRightLeafNode() != null)
+				rightCover = cover(root.getRightLeafNode(), n);
+			
+			return (leftCover || rightCover);
+		}
+		
+//		return false;
+	}
+
+	public boolean isSameTree(TreeNode p, TreeNode q) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 583-588 into a separate method.